eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -542,89 +549,187 @@ instance Semigroup ScriptResult where
+debugPlutus ::
+  forall (l :: Language).
+  (Typeable l, IsLanguage l) =>
+  Version ->
+  String ->
+  PlutusDebugInfo l
 debugPlutus version db =
   case B64.decode (BSU.fromString db) of
     Left e -> DebugBadHex (show e)
     Right bs ->
-      case decodeFull' version bs of
+      case plutusDebugLang <$> (decodeFull' version bs :: Either DecoderError (PlutusDebug l)) of
         Left e -> DebugCannotDecode (show e)
-        Right pdb@(PlutusDebugV1 cm units script ds pv) ->
-          case PV1.evaluateScriptRestricting
-            (transProtocolVersion pv)
-            PV1.Verbose
-            (getEvaluationContext cm)
-            (transExUnits units)
-            script
-            ds of
-            (logs, Left e) -> DebugInfo logs (PlutusErrorV1 e) pdb
-            (_, Right ex) -> DebugSuccess ex
-        Right pdb@(PlutusDebugV2 cm units script ds pv) ->
-          case PV2.evaluateScriptRestricting
-            (transProtocolVersion pv)
-            PV2.Verbose
-            (getEvaluationContext cm)
-            (transExUnits units)
-            script
-            ds of
-            (logs, Left e) -> DebugInfo logs (PlutusErrorV2 e) pdb
-            (_, Right ex) -> DebugSuccess ex
+        Right (sl, PlutusDebug costModel exUnits scripts pData protVer) ->
+          let evalV1 =
+                PV1.evaluateScriptRestricting
+                  (transProtocolVersion protVer)
+                  PV1.Verbose
+                  (getEvaluationContext costModel)
+                  (transExUnits exUnits)
+                  scripts
+              evalV2 =
+                PV2.evaluateScriptRestricting
+                  (transProtocolVersion protVer)
+                  PV2.Verbose
+                  (getEvaluationContext costModel)
+                  (transExUnits exUnits)
+                  scripts
+           in case sl of
+                SPlutusV1 ->
+                  case evalV1 $ unDataV1 pData of
+                    (logs, Left e) ->
+                      DebugInfo
+                        logs
+                        (PlutusErrorV1 e)
+                        ( mkPlutusDebug
+                            isLanguage
+                            costModel
+                            exUnits
+                            scripts
+                            (unDataV1 pData)
+                            protVer
+                        )
+                    (_, Right ex) -> DebugSuccess ex
+                SPlutusV2 ->
+                  case evalV2 $ unDataV2 pData of
+                    (logs, Left e) ->
+                      DebugInfo
+                        logs
+                        (PlutusErrorV2 e)
+                        ( mkPlutusDebug
+                            isLanguage
+                            costModel
+                            exUnits
+                            scripts
+                            (unDataV2 pData)
+                            protVer
+                        )
```

Comment:
Reduce duplication and use the proper `SLanguage` type:

```suggestion
                plututsDebug = mkPlutusDebug sl costModel exUnits scripts pData protVer
           in case sl of
                SPlutusV1 ->
                  case evalV1 $ unDataV1 pData of
                    (logs, Left e) -> DebugInfo logs (PlutusErrorV1 e) plutusDebug
                    (_, Right ex) -> DebugSuccess ex
                SPlutusV2 ->
                  case evalV2 $ unDataV2 pData of
                    (logs, Left e) -> DebugInfo logs (PlutusErrorV2 e) plutusDebug
```